File: tools/Generator/Amalgamator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace Generator
{
    class HeaderFile : IComparable<HeaderFile>
    {
        public String filePath; // complete path
        public String fileBuff; // file content without #include "xxx.h"
        public List<String> incList; // include list with complete path

        public HeaderFile()
        {
            filePath = "";
            fileBuff = "";
            incList = new List<String>();
        }

        public int CompareTo(HeaderFile hf) // require for sorting...
        {
            if (hf.incList.Count > incList.Count)
            {
                return -1;
            }
            else if (hf.incList.Count == incList.Count)
            {
                return 0;
            }
            else
            {
                return 1;
            }
        }
    }

    public class Amalgamator
    {
        static int lineCount;
        static HeaderFile ReadHeaderFile(string path)
        {
            Console.Out.WriteLine("Processing: " + path);

            HeaderFile headerFile = new HeaderFile();
            headerFile.filePath = path;

            StreamReader sr = new StreamReader(path);

            Regex regex = new Regex("#include *\"(_*/*.*[0-9]*[A-Z]*[a-z]*)+\""); // file name can contain _/.number or char
            Regex regex2 = new Regex("\"(_*/*.*[0-9]*[A-Z]*[a-z]*)+\"");

            String curLine;
            while ((curLine = sr.ReadLine()) != null)
            {
                Match m = regex.Match(curLine);
                if (m.Success)
                {
                    String includeSection = curLine.Substring(m.Index, m.Length);
                    m = regex2.Match(includeSection);
                    String relativeHeaderPath = includeSection.Substring(m.Index, m.Length).Substring(1, m.Length - 2);

                    String curDir = Path.GetDirectoryName(path);
                    String incFilePath = Path.GetFullPath(curDir + "/" + relativeHeaderPath);
                    if (File.Exists(incFilePath))
                    {
                        headerFile.incList.Add(incFilePath);
                    }
                    else
                    {
                        Console.Out.WriteLine("Error: missing " + incFilePath);
                        return null;
                    }
                }
                else
                {
                    lineCount++;
                    headerFile.fileBuff += "\r\n" + curLine;
                }
            }

            sr.Close();
            return headerFile;
        }

        static List<HeaderFile> SerachForHeaderFiles(string sourceDir)
        {
            string[] fileEntries = Directory.GetFiles(Path.GetFullPath(sourceDir), "*.h", SearchOption.AllDirectories);

            List<HeaderFile> headerFileList = new List<HeaderFile>();

            foreach (string fileName in fileEntries)
            {
                HeaderFile hf = ReadHeaderFile(fileName);
                if (hf != null)
                {
                    headerFileList.Add(hf);
                }
                else
                {
                    return null;
                }
            }

            return headerFileList;
        }

        static bool IsRequiredHeadersWritten(HeaderFile headerFile, List<HeaderFile> WrittenFiles)
        {
            bool allWritten = true;
            for (int i = 0; i < headerFile.incList.Count; i++)
            {
                bool curFileFound = false;
                for (int j = 0; j < WrittenFiles.Count; j++)
                {
                    if (WrittenFiles[j].filePath.CompareTo(headerFile.incList[i]) == 0)
                    {
                        curFileFound = true;
                        break;
                    }
                }
                if (!curFileFound)
                {
                    allWritten = false;
                    break;
                }
            }
            return allWritten;
        }

        static bool AmalgamateHeaderFiles(String frameworkPath, String frameworkName)
        {
            List<HeaderFile> headerFileList = SerachForHeaderFiles(frameworkPath);

            if (headerFileList == null)
                return false;

            headerFileList.Sort();

            String outHeaderFile = "\r\n// ========== RFC Generator v1.0 - "+ DateTime.Now.ToString("yyyy-MM-dd HH:mm tt") + " ==========\r\n";

            String tmp = "_" + frameworkName.Replace(' ', '_').ToUpper();
            outHeaderFile += "\r\n#ifndef " + tmp + "_H_\r\n#define " + tmp + "_H_ \r\n\r\n#define AMALGAMATED_VERSION\r\n";

            List<HeaderFile> WrittenFiles = new List<HeaderFile>();

            Console.Out.WriteLine("\n");

            // get 0 includes
            foreach (HeaderFile headerFile in headerFileList)
            {
                if (headerFile.incList.Count == 0)
                {
                    Console.Out.WriteLine("Adding: " + headerFile.filePath);
                    outHeaderFile += "\r\n\r\n// =========== " + Path.GetFileName(headerFile.filePath) + " ===========" + headerFile.fileBuff;
                    WrittenFiles.Add(headerFile);
                }
                else
                {
                    break;
                }
            }

            if (WrittenFiles.Count != 0) // no circular dependancies
            {
                // remove 0 includes
                foreach (HeaderFile headerFile in WrittenFiles)
                {
                    headerFileList.Remove(headerFile);
                }

                while (headerFileList.Count != 0)
                {
                    foreach (HeaderFile headerFile in headerFileList)
                    {
                        if (headerFile.incList.Count <= WrittenFiles.Count)
                        {
                            if (IsRequiredHeadersWritten(headerFile, WrittenFiles))
                            {
                                Console.Out.WriteLine("Adding: " + headerFile.filePath);
                                outHeaderFile += "\r\n\r\n// =========== " + Path.GetFileName(headerFile.filePath) + " ===========" + headerFile.fileBuff;
                                WrittenFiles.Add(headerFile);
                            }
                        }
                        else
                        {
                            break;
                        }
                    }

                    // remove written files from the list
                    foreach (HeaderFile headerFile in WrittenFiles)
                    {
                        headerFileList.Remove(headerFile);
                    }
                }

                outHeaderFile += "\r\n\r\n#endif\r\n\r\n";

                Console.Out.WriteLine("\nWriting: " + frameworkName + ".h lines: " + lineCount + "\n");
                StreamWriter sw = new StreamWriter(frameworkName + ".h");
                sw.Write(outHeaderFile);
                sw.Close();

                return true;
            }
            else
            {
                Console.Out.WriteLine("\nError: Circular dependancy detected!");
                return false;
            }
        }

        public static bool MakeAmalgamatedRFC(string frameworkPath, string outputName)
        {
            if(AmalgamateHeaderFiles(frameworkPath, outputName))
            {
                AmalgamateSourceFiles(frameworkPath, outputName);
                return true;
            }
            return false;
        }

        static void AmalgamateSourceFiles(String frameworkPath, String frameworkName)
        {
            string[] fileEntries = Directory.GetFiles(frameworkPath, "*.cpp", SearchOption.AllDirectories);

            String outSourceFile = "\r\n// ========== RFC Generator v1.0 ==========\r\n";

            outSourceFile += "\r\n\r\n#include \"" + frameworkName + ".h\"\r\n";

            int lineCount = 0;

            foreach (string fileName in fileEntries)
            {
                Console.Out.WriteLine("Adding: " + fileName);

                StreamReader sr = new StreamReader(fileName);

                outSourceFile += "\r\n\r\n// =========== " + Path.GetFileName(fileName) + " ===========";

                Regex regex = new Regex("#include *\"(_*/*.*[0-9]*[A-Z]*[a-z]*)+\""); // file name can contain _/.number or char

                String curLine;
                while ((curLine = sr.ReadLine()) != null)
                {
                    lineCount++;
                    Match m = regex.Match(curLine);
                    if (!m.Success) // bypass #includes...
                    {
                        outSourceFile += "\r\n" + curLine;
                    }
                }

                sr.Close();
            }

            Console.Out.WriteLine("\nWriting: " + frameworkName + ".cpp lines: " + lineCount + "\n");
            StreamWriter sw = new StreamWriter(frameworkName + ".cpp");
            sw.Write(outSourceFile);
            sw.Close();

        }
    }
}

File: tools/rfc-builder/amalgamator.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"
)

var includeRegex = regexp.MustCompile(`#include\s*"([^"]+)"`)

// HeaderFile holds the representation of a parsed C++ header file
type HeaderFile struct {
	FilePath string   // complete path
	FileBuff string   // file content without local includes
	IncList  []string // resolved absolute paths of local includes
}

// ReadHeaderFile parses a header file and extracts its local includes
func ReadHeaderFile(path string, logger io.Writer) (*HeaderFile, error) {
	fmt.Fprintln(logger, "Processing: "+path)

	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	hf := &HeaderFile{
		FilePath: path,
		IncList:  []string{},
	}

	scanner := bufio.NewScanner(file)
	curDir := filepath.Dir(path)

	for scanner.Scan() {
		line := scanner.Text()
		matches := includeRegex.FindStringSubmatch(line)
		if len(matches) > 1 {
			relPath := matches[1]
			// Resolve path and clean it up
			incFilePath := filepath.Clean(filepath.Join(curDir, relPath))

			// Check if file exists
			if _, err := os.Stat(incFilePath); err == nil {
				hf.IncList = append(hf.IncList, incFilePath)
			} else {
				fmt.Fprintln(logger, "Error: missing "+incFilePath)
				return nil, fmt.Errorf("missing header: %s", incFilePath)
			}
		} else {
			hf.FileBuff += "\r\n" + line
		}
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return hf, nil
}

// SearchForHeaderFiles scans a directory for all .h files recursively
func SearchForHeaderFiles(sourceDir string, logger io.Writer) ([]*HeaderFile, error) {
	var headers []*HeaderFile

	err := filepath.Walk(sourceDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if !info.IsDir() && filepath.Ext(path) == ".h" {
			hf, err := ReadHeaderFile(path, logger)
			if err != nil {
				return err
			}
			headers = append(headers, hf)
		}
		return nil
	})

	if err != nil {
		return nil, err
	}

	return headers, nil
}

// isRequiredHeadersWritten checks if all headers in incList are already written
func isRequiredHeadersWritten(incList []string, writtenFiles map[string]bool) bool {
	for _, inc := range incList {
		if !writtenFiles[inc] {
			return false
		}
	}
	return true
}

// AmalgamateHeaderFiles performs dependency-ordered amalgamation on header files
func AmalgamateHeaderFiles(frameworkPath string, frameworkName string, outputDir string, logger io.Writer) (int, error) {
	headers, err := SearchForHeaderFiles(frameworkPath, logger)
	if err != nil {
		return 0, err
	}

	if len(headers) == 0 {
		return 0, fmt.Errorf("no header files found in %s", frameworkPath)
	}

	// Sort headers ascending by the length of their dependencies
	sort.Slice(headers, func(i, j int) bool {
		return len(headers[i].IncList) < len(headers[j].IncList)
	})

	// Format timestamp in yyyy-MM-dd hh:mm tt format (e.g. 2026-06-19 08:52 AM)
	timestamp := time.Now().Format("2006-01-02 03:04 PM")
	outHeaderFile := "\r\n// ========== RFC Generator v1.0 - " + timestamp + " ==========\r\n"

	guardName := "_" + strings.ToUpper(strings.ReplaceAll(frameworkName, " ", "_"))
	outHeaderFile += "\r\n#ifndef " + guardName + "_H_\r\n#define " + guardName + "_H_ \r\n\r\n#define AMALGAMATED_VERSION\r\n"

	writtenMap := make(map[string]bool)
	var writtenList []*HeaderFile
	var remaining []*HeaderFile

	// 1. Process 0-dependency headers
	for _, hf := range headers {
		if len(hf.IncList) == 0 {
			fmt.Fprintln(logger, "Adding: "+hf.FilePath)
			outHeaderFile += "\r\n\r\n// =========== " + filepath.Base(hf.FilePath) + " ===========" + hf.FileBuff
			writtenMap[hf.FilePath] = true
			writtenList = append(writtenList, hf)
		} else {
			remaining = append(remaining, hf)
		}
	}

	if len(writtenList) == 0 {
		fmt.Fprintln(logger, "\nError: Circular dependency detected (no base files with 0 includes)!")
		return 0, fmt.Errorf("circular dependency detected (no files with 0 includes)")
	}

	// 2. Process remaining headers with dependency ordering
	for len(remaining) > 0 {
		progress := false
		var nextRemaining []*HeaderFile

		for _, hf := range remaining {
			if len(hf.IncList) <= len(writtenMap) && isRequiredHeadersWritten(hf.IncList, writtenMap) {
				fmt.Fprintln(logger, "Adding: "+hf.FilePath)
				outHeaderFile += "\r\n\r\n// =========== " + filepath.Base(hf.FilePath) + " ===========" + hf.FileBuff
				writtenMap[hf.FilePath] = true
				progress = true
			} else {
				nextRemaining = append(nextRemaining, hf)
			}
		}

		if !progress {
			fmt.Fprintln(logger, "\nError: Circular dependency detected among remaining headers!")
			return 0, fmt.Errorf("circular dependency detected among remaining headers")
		}
		remaining = nextRemaining
	}

	outHeaderFile += "\r\n\r\n#endif\r\n\r\n"

	// Count total lines in the amalgamated buffer
	lineCount := strings.Count(outHeaderFile, "\n") + 1

	fmt.Fprintf(logger, "\nWriting: %s.h lines: %d\n", frameworkName, lineCount)
	
	outPath := filepath.Join(outputDir, frameworkName+".h")
	err = os.WriteFile(outPath, []byte(outHeaderFile), 0644)
	if err != nil {
		return 0, err
	}

	return lineCount, nil
}

// AmalgamateSourceFiles concatenates source files skipping local header includes
func AmalgamateSourceFiles(frameworkPath string, frameworkName string, outputDir string, logger io.Writer) error {
	var cppFiles []string
	err := filepath.Walk(frameworkPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if !info.IsDir() && filepath.Ext(path) == ".cpp" {
			cppFiles = append(cppFiles, path)
		}
		return nil
	})
	if err != nil {
		return err
	}

	outSourceFile := "\r\n// ========== RFC Generator v1.0 ==========\r\n"
	outSourceFile += "\r\n\r\n#include \"" + frameworkName + ".h\"\r\n"

	var lineCount int
	for _, cppFile := range cppFiles {
		fmt.Fprintln(logger, "Adding: "+cppFile)
		outSourceFile += "\r\n\r\n// =========== " + filepath.Base(cppFile) + " ==========="

		file, err := os.Open(cppFile)
		if err != nil {
			return err
		}

		scanner := bufio.NewScanner(file)
		for scanner.Scan() {
			line := scanner.Text()
			lineCount++
			if !includeRegex.MatchString(line) {
				outSourceFile += "\r\n" + line
			}
		}
		file.Close()
		if err := scanner.Err(); err != nil {
			return err
		}
	}

	fmt.Fprintf(logger, "\nWriting: %s.cpp lines: %d\n\n", frameworkName, lineCount)

	outPath := filepath.Join(outputDir, frameworkName+".cpp")
	return os.WriteFile(outPath, []byte(outSourceFile), 0644)
}

// MakeAmalgamatedRFC is the entry point for both header and source amalgamation
func MakeAmalgamatedRFC(frameworkPath string, outputName string, outputDir string, logger io.Writer) (bool, error) {
	_, err := AmalgamateHeaderFiles(frameworkPath, outputName, outputDir, logger)
	if err != nil {
		return false, err
	}

	err = AmalgamateSourceFiles(frameworkPath, outputName, outputDir, logger)
	if err != nil {
		return false, err
	}

	return true, nil
}
